### PR TITLE
Add HuBERT code and features

### PR DESCRIPTION
**Pull request by StellaYuxiao**
If it's your first time running it, you may need:
!pip install transformers torchaudio librosa matplotlib seaborn scikit-learn pandas

### CODE REVIEW

**Review by mifnd — CHANGES_REQUESTED** [review]
Nice! This is a great start!😃
It looks like this is a Jupyter notebook file that has been converted to a .py-file. Using Jupiter is fine for testing on your own computer. However, in our GitHub repo we need only python scripts that can be easily run by all group members. 

Thus to get there:
- Split up in separate scripts (one for making Hubert features, one for each of the data visualizations)
- Place the scripts in the appropriate folders
- Read and write datasets to appropriate folders
- You mention that certain libraries may need to one installed. These should be added to the environment.yml-file. 
- I don't think we need the html-file. Isn't this just a print of the Jupyter-file?

**Comment by SWalles** [comment]
I have moved the files to appropriate directories and updated the environment.yml file. 
Please use command:
<conda env create -f environment.yml> to create a new conda environment
OR
<conda env update --file environment.yml  --prune> to update an existing environment

I can see you have put a lot of effort into this. Thank you!

To mostly elaborate on things Mikkel also noted:
- Please use the pathlib library for working with file paths (instead of os)
- Please do not specify absolute paths, but use paths relative to the base "compsci760_voice_diagnosis" folder (you can look at some of the other code on the github for examples)

File: data/processed/hubert_features.csv
Comment (SWalles) [inline data/processed/hubert_features.csv:1]: Running the code creates a file called "hubert_self_attentive_features.csv" is that the same as this file just a different name? If so can we either change the name in the code or the file name?

Comment (SWalles) [reply data/processed/hubert_features.csv:1]: Is it the same as "hubert_features.csv" I mean?

**Comment by StellaYuxiao** [comment]
Not really, “Hubert_features.csv” is using the mean pooling method, the
other one is using the new self-attentive pooling method, it’s a new
feature data, thank you!

SWalles ***@***.***>于2025年5月27日 周二18:06写道：

**Review by SWalles — APPROVED** [review]
Thank you. This looks good

File: code/modeling/HuBERT/HuBERT.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+model.eval()
+
+# Self-attentive pooling
+def self_attentive_pooling(hidden_states):
+    weights = torch.nn.functional.softmax(hidden_states.mean(dim=-1), dim=1).unsqueeze(-1)
+    return (hidden_states * weights).sum(dim=1)
+
+# Collect data
+data = []
+sound_types = ["Healthy", "Laryngocele", "Vox_senilis"]
```
Comment (mifnd) [inline code/modeling/HuBERT/HuBERT.py:26]: Note: These labels are different from what we use in the rest of the project...

File: environment.yml [hunk 2]
```diff
@@ -9,4 +9,9 @@ dependencies:
   - scikit-learn=1.6.1
   - pysoundfile=0.13.1
   - libsndfile=1.2.2
-
+  - xgboost=2.1.4
+  - matplotlib
+  - seaborn
+  - pytorch
+  - torchaudio
+  - transformers
```
Comment (mifnd) [inline environment.yml:17]: I think you need to merge the main branch into your branch

**Comment by mifnd** [comment]
I have updated the PR:
- Merged main into your branch
- Moved the script to the feature extraction folder
- Renamed to script to HuBERT_self_attentive.py, as it seems like the self attentive method is used in the script.

Other than that I think it looks really good! If you agree with my edits, then let me know, and I will merge your PR.

**Review by mifnd — APPROVED** [review]
